Allow source files to be found system-wide and don't hardcode /usr prefix on UNIX

chewi opened the pull request:
`/usr/share/{NCINE_LINUX_PACKAGE}/Source` is tried before `~/.local/share/{NCINE_LINUX_PACKAGE}/Source`, which is slightly counter-intuitive but necessary. If you favour the user directory, it will be automatically created on first launch and the system-wide directory will never be considered again, even if the user directory is still empty. The game shouldn't try to create the system-wide directory because it wouldn't have permission to do so.

deathkiller commented:
Thanks for contribution!
- It seems AppImage build is broken because of changes in `InitializePaths()`.
It looks like you simplified the `NCINE_PACKAGED_CONTENT_PATH` section too much. I'd rather leave it the way it was, because it's special flag for AppImages and splitting the section into 2 `#ifs` makes the code less readable for me.
- I'm a bit afraid to merge these changes, because some parts are shared across all supported platforms and Linux distributions (and files in `Shared` directory are shared across all my projects).
- Do you know what values `CMAKE_INSTALL_PREFIX` can have on different systems? Is it set to `"/usr/"` everywhere by default? Is `CMAKE_INSTALL_PREFIX` needed also in `"/lib/os-release"`?
- I'm not sure about prioritizing the system-wide `Source` directory over the user's, because then you would have to uninstall the system-wide `Source` first to be able to provide your own (e.g. if you want to test different versions). I guess I'll have to think about it a little more.

chewi commented:
> It seems AppImage build is broken because of changes in `InitializePaths()`. It looks like you simplified the `NCINE_PACKAGED_CONTENT_PATH` section too much. I'd rather leave it the way it was, because it's special flag for AppImages and splitting the section into 2 `#ifs` makes the code less readable for me.

Okay, I dropped that commit. It wasn't essential.

> I'm a bit afraid to merge these changes, because some parts are shared across all supported platforms and Linux distributions (and files in `Shared` directory are shared across all my projects).

I'm not sure what you mean. Apart from the `target_compile_definitions` line, these changes are all under `DEATH_TARGET_UNIX`. I think they make sense for any Unix-like platform.

> Do you know what values `CMAKE_INSTALL_PREFIX` can have on different systems? Is it set to `"/usr/"` everywhere by default? Is `CMAKE_INSTALL_PREFIX` needed also in `"/lib/os-release"`?

Actually, the default set by CMake is `/usr/local`, which means that as things stand right now, the game looks in the wrong place for `Content` unless you explicitly pass `-DCMAKE_INSTALL_PREFIX=/usr` to CMake. Distribution packages will do this anyway, but users building it manually probably won't. This change is therefore a good thing. I made this change to support Gentoo Prefix, which installs packages to arbitrary locations like `/foo/bar/usr`.

Good point about `os-release` though. Using `CMAKE_INSTALL_PREFIX` here would make sense for Gentoo Prefix, but probably not for anything else. I've reverted that bit.

> I'm not sure about prioritizing the system-wide `Source` directory over the user's, because then you would have to uninstall the system-wide `Source` first to be able to provide your own (e.g. if you want to test different versions). I guess I'll have to think about it a little more.

Yeah, I didn't really want to do it that way, but I couldn't think of an alternative without dropping the automatic directory creation. Even then, users who have already run the game once without any files won't see any system-wide files installed subsequently. It would need to check whether the directory was empty or not. You don't currently have a helper for checking directory contents. If I was to start writing a new game, I'd probably use something like [PhysicsFS](https://icculus.org/physfs/), but I wasn't going to go that far here. :smile:

deathkiller commented:
I think I can change it to check presence of `"Anims.j2a"` or `"AnimsSw.j2a"` in the directory in `InitializePaths()`, something like this:
https://github.com/deathkiller/jazz2-native/blob/bdf0ebddf35776dbed6f88a06c181782ce12d72c/Sources/Main.cpp#L1009-L1012

Then the priority would be as expected (user first).
Downside is that it would have to iterate through all files in the directory twice, because Linux is usually case sensitive. I'm not sure how much it would slow down the startup, hopefully not much.

> I'm not sure what you mean. Apart from the target_compile_definitions line, these changes are all under DEATH_TARGET_UNIX. I think they make sense for any Unix-like platform.

I meant mainly FreeBSD which I know is not very popular, but the game works there.

> Actually, the default set by CMake is /usr/local, which means that as things stand right now, the game looks in the wrong place for Content unless you explicitly pass -DCMAKE_INSTALL_PREFIX=/usr to CMake. Distribution packages will do this anyway...

If you think nobody will complain, then it's fine for me.

I looked at PhysicsFS, but I didn't like it, I don't remember why and I wanted migrate C# code as fast as possible, so I created my own IO API. The advantage is that I can extend it as much as I want.
